pkg/shares/share_merging.go

```diff
@@ -132,3 +134,43 @@ func ParseMsgs(shares [][]byte) (coretypes.Messages, error) {
 		MessagesList: msgList,
 	}, nil
 }
+
+// ShareSequence represents a contiguous sequence of shares that are part of the
+// same namespace and message. For compact shares, one share sequence exists per
+// reserved namespace. For sparse shares, one share sequence exists per message.
+type ShareSequence struct {
+	NamespaceID namespace.ID
+	Shares      []Share
+}
+
+func ParseShares(rawShares [][]byte) (result []ShareSequence, err error) {
+	currentSequence := ShareSequence{}
+
+	for _, rawShare := range rawShares {
+		share, err := NewShare(rawShare)
+		if err != nil {
+			return result, err
+		}
+		infoByte, err := share.InfoByte()
+		if err != nil {
+			return result, err
+		}
+		if infoByte.IsMessageStart() {
+			if len(currentSequence.Shares) > 0 {
+				result = append(result, currentSequence)
+			}
+			currentSequence = ShareSequence{
+				Shares:      []Share{share},
+				NamespaceID: share.NamespaceID(),
+			}
+		}
+		if !infoByte.IsMessageStart() {
```

Comment:
[nit][optional]
could we get rid of this second check or using an else?

Reply:
Resolved in https://github.com/celestiaorg/celestia-app/pull/828/commits/16f8d623230f56a396ae00dd02bae9b16f50f0b5